src/core/textlayout.js

```diff
@@ -0,0 +1,176 @@
+/* -*- tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set shiftwidth=2 tabstop=2 autoindent cindent expandtab: */
+/* Copyright 2012 Mozilla Foundation 
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License. 
+ */
+/* globals assert, ColorSpace, DecodeStream, Dict, Encodings,
+           error, ErrorFont, Font, FONT_IDENTITY_MATRIX, fontCharsToUnicode,
+           FontFlags, ImageKind, info, isArray, isCmd, isDict, isEOF, isName,
+           isNum, isStream, isString, JpegStream, Lexer, Metrics,
+           Name, Parser, Pattern, PDFJS, warn, Util, Promise,
+           createPromiseCapability, QuadTree */
+ 
+'use strict';
+ 
+var TextLayoutEvaluator = (function TextLayoutEvaluatorClosure() {
+  function TextLayoutEvaluator() {
+  }
+ 
+  var NonWhitespaceRegexp = /\S/;
+  function isAllWhitespace(str) {
+    return !NonWhitespaceRegexp.test(str);
+  }
+
+  // Trying to minimize Date.now() usage and check every 100 time 
+  var TIME_SLOT_DURATION_MS = 20;
+  var CHECK_TIME_EVERY = 100;
+  function TimeSlotManager() {
+    this.reset();
+  }
+  TimeSlotManager.prototype = {
+    check: function TimeSlotManager_check() {
+      if (++this.checked < CHECK_TIME_EVERY) {
+        return false;
+      }
+      this.checked = 0;
+      return this.endTime <= Date.now();
+    },
+    reset: function TimeSlotManager_reset() {
+      this.endTime = Date.now() + TIME_SLOT_DURATION_MS;
+      this.checked = 0;
+    }
+  };
+  var deferred = Promise.resolve();
+
+  TextLayoutEvaluator.prototype = {
+    overlapBy : function (min1, min2, max1, max2, by) {
```

Comment:
Nit: as outlined [here](https://github.com/mozilla/pdf.js/wiki/Style-Guide#classes), please make sure that you name all anonymous functions like this
```javascript
overlapBy: function TextLayoutEvaluator_overlapBy(...) {
```
This applies to functions below as well.